corporate/lib/stripe.py

```diff
@@ -682,15 +694,22 @@ def update_or_create_customer(
             )
             return customer
 
+    @override
+    def do_change_plan_type(self, plan_type: Optional[int] = None) -> None:
+        from zerver.actions.realm_settings import do_change_realm_plan_type
+
+        if plan_type is None:
+            plan_type = Realm.PLAN_TYPE_STANDARD
+        do_change_realm_plan_type(self.realm, plan_type, acting_user=self.user)
```

Comment:
I don't like having this default value a bit hidden in the implementation, but I didn't have a better idea in the moment. The default will need to be different for the realm billing session vs the remote server billing session.

Reply:
I think structurally, we probably need the caller to be passing the plan type into this function.

Reply:
I made the following tweak here:
```diff
diff --git a/corporate/lib/stripe.py b/corporate/lib/stripe.py
index bb19a733d2..78b551b206 100644
--- a/corporate/lib/stripe.py
+++ b/corporate/lib/stripe.py
@@ -402,7 +402,7 @@ class BillingSession(ABC):
         pass
 
     @abstractmethod
-    def do_change_plan_type(self, plan_type: Optional[int] = None) -> None:
+    def do_change_plan_type(self, customer_plan_type: int = None) -> None:
         pass
 
     @abstractmethod
@@ -693,11 +693,19 @@ class RealmBillingSession(BillingSession):
             return customer
 
     @override
-    def do_change_plan_type(self, plan_type: Optional[int] = None) -> None:
+    def do_change_plan_type(self, tier: Optional[int], is_sponsored=False) -> None:
         from zerver.actions.realm_settings import do_change_realm_plan_type
 
-        if plan_type is None:
+        # This function needs to translate between the different
+        # formats of CustomerPlan.tier and Realm.plan_type.
+        if is_sponsored:
+            plan_type = Realm.PLAN_TYPE_STANDARD_FREE
+        elif tier == CustomerPlan.STANDARD:
             plan_type = Realm.PLAN_TYPE_STANDARD
+        elif tier == CustomerPlan.PLUS:
+            plan_type = Realm.PLAN_TYPE_PLUS
+        else:
+            raise AssertionError("Unexpected tier")
         do_change_realm_plan_type(self.realm, plan_type, acting_user=self.user)
 
     @override
@@ -707,7 +715,7 @@ class RealmBillingSession(BillingSession):
 
         from zerver.actions.message_send import internal_send_private_message
 
-        self.do_change_plan_type(Realm.PLAN_TYPE_STANDARD_FREE)
+        self.do_change_plan_type(tier=None, is_sponsored=True)
         customer = self.get_customer()
         if customer is not None and customer.sponsorship_pending:
             customer.sponsorship_pending = False
@@ -1100,7 +1108,7 @@ def process_initial_upgrade(
         )
         stripe.Invoice.finalize_invoice(stripe_invoice)
 
-    billing_session.do_change_plan_type()
+    billing_session.do_change_plan_type(tier=plan_tier)
 
 
 def update_license_ledger_for_manual_plan(
```